### DISCUSSION TITLE
Passing a parameter to a custom hook

### DISCUSSION THREAD

**Discussion by leandroromano** [directus/directus#16967] [Feature Requests]
Hi everyone!

I'm trying to pass a parameter/variable to my custom hook, specifically to the method filter like this:

```typescript
import { defineHook } from '@directus/extensions-sdk'

export default defineHook(({ filter, action, init }) => {
  filter('*.read', async (data: any, { query }, context) => {
    if (context.customParameters.language) {
      //filter query to translate to specific language
    }
  })
})
```

I was trying to pass parameters through context object and also from the meta object but i couldn't achieve it. Also i noticed that i can pass a query param or graphql variable and that variable is able to get from the init method middlewares.after but i need the variable on the filter method just to filter the queries. So my doubt is if it is possible to pass a parameter to the filter method?

Thanks in advance and regards!

**Comment by br41nslug** [comment]
from where are you trying to pass a variable?

**Reply by leandroromano** [reply]
I would like to pass a variable like a query param or in the body of the request. So it would be good if is available the object `request` on the method `filter` to the event `read`. What do you think, is possible?

**Reply by br41nslug** [reply]
Hmmm interesting, most of the unknown/unwanted incoming query parameters get sanitized pretty heavily. It is a bit of a workaround and suboptimal but the only method i can currently think of is to add a middleware (using `init('middleware.after')`) that would add your property to the `request.sanitizedQuery` as that is the object eventually passed to the `filter`. Not the nicest way but i think it should work. Beware that not all `items.read` events are associated with a request, for instance reads done by Flows, Custom Hooks or Endpoints.

**Reply by leandroromano** [reply]
Thank you for your response @br41nslug ! I tried that way and it worked for `items.read` triggered by the endpoint to query some item, but unfortunately it didn't work when triggering the `items.read` from a graphql query using the endpoint `/graphql`. So, is there a way to make it work from a graphql query? Maybe extending or doing a workaround on the package and making use of [patch-package](https://github.com/ds300/patch-package)

**Reply by br41nslug** [reply]
Ah i don't think this will work for graphql but you'd have to update `res.locals.graphqlParams` ([source](https://github.com/directus/directus/blob/c3ecbed0b4143452ae5e0cd9196d6c4a53d57b0b/api/src/controllers/graphql.ts#L36)) instead from the middleware  as thats the info passed to the service but graphql is quite particular in its data so this might break things.
When doing these sorts of trickery would recommend to stick with the REST API to avoid unneeded complexity.

**Reply by leandroromano** [reply]
I tried this way and it seems to work but i have the doubt if this could break anything, what do you think?

```typescript
import { defineHook } from '@directus/extensions-sdk'

export default defineHook(({ filter, action, init }, { services }) => {
  filter('*.read', async (data: any, { query }, context) => {
    console.log(context.accountability?.customParams.languages)
    console.log(context.accountability?.customParams.fields)
    return data
  })

  init('middlewares.after', ({ app }) => {
    app.use((_req, res, next) => {
      // Modify accountability object
      const languages = ['es', 'en', 'fr']
      const fields = ['name', 'description']
      _req.accountability = {
        ..._req.accountability,
        customParams: { languages, fields }
      }
      next()
    })
  })
})
```

I'm able to read this customParams from `context.accountability` object

Thanks again 🙌🏻

**Reply by br41nslug** [reply]
Oh `context.accountability` is a good one, didnt think of that. No i dont think that should break anything.

**Reply by leandroromano** [reply]
Amazing! So definitely i'm going to use this way to pass a parameter to the filter `items.read`. Thank you so much for your time!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by jperestrelo** [comment]
> Hmmm interesting, most of the unknown/unwanted incoming query parameters get sanitized pretty heavily. It is a bit of a workaround and suboptimal but the only method i can currently think of is to add a middleware (using `init('middleware.after')`) that would add your property to the `request.sanitizedQuery` as that is the object eventually passed to the `filter`. Not the nicest way but i think it should work. Beware that not all `items.read` events are associated with a request, for instance reads done by Flows, Custom Hooks or Endpoints.

Hi @br41nslug, I also need to pass some custom parameter so I can use in my filter function, I followed your suggestion here and I can get it on the event object, so all good. I'm just afraid that I'm not doing it correctly, because since sanitzedQuery was not extensible I had to clone it. 
Here is my code:

```
 init("middlewares.after", ({ app }) => {
        app.use((req, res, next) => {

            const customParamValue = req.query.customParam;

            let modifiedQuery = { ...req.sanitizedQuery };
            modifiedQuery.customParam = customParamValue;
            req.sanitizedQuery = modifiedQuery;

            next();
        });
    });
```


I'm really not sure if this is the correct way, can you please give some guidance?
Thanks a lot!

**Reply by br41nslug** [reply]
I literally called it "a workaround and suboptimal" so the correct way this will never be. But have a look at the other comments for inspiration https://github.com/directus/directus/discussions/16967#discussioncomment-4571887

**Reply by jperestrelo** [reply]
thanks @br41nslug but meanwhile, I solved my need with a custom endpoint. I was going to delete this post but forgot. Thanks anyway for taking the time to answer.
